storage: add fast-path for ticking quiesced/dormant replicas

petermattis opened the pull request:
Add fast-path for ticking quiesced/dormant replicas which avoids going
through the Raft scheduler and avoids grabbing Replica.raftMu which can be
held for significant periods of time.

Fixes #17609

petermattis committed:
storage: add fast-path for ticking quiesced/dormant replicas

Add fast-path for ticking quiesced/dormant replicas which avoids going
through the Raft scheduler and avoids grabbing Replica.raftMu which can
be held for significant periods of time.

Fixes #17609

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17617)
<!-- Reviewable:end -->

petermattis commented:
@bdarnell If there are caveats to changing the Raft tick settings as is done by the first commit, I'll document the rationale behind the current settings thoroughly.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I described the issues with changing the raft values in https://github.com/cockroachdb/cockroach/issues/17609#issuecomment-322207809. I don't think we've done any experimentation to show whether the coarser randomization of election timeouts makes a difference in practice. 

I'd rather just do the fast-path for now and leave the constants alone.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/base/config.go, line 77 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-KrWCEaSzreyBWcjBZjq:-KrWCEaSzreyBWcjBZjr:b-kqmzvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/36f9837c96863fde717b927fe908612607fc9334/pkg/base/config.go#L77)):*
> ```Go
> 
> var defaultRaftElectionTimeoutTicks = envutil.EnvOrDefaultInt(
> 	"COCKROACH_RAFT_ELECTION_TIMEOUT_TICKS", 3)
> ```

This change is backwards-incompatible for anyone who might have been using this environment variable. (Have we recommended this to anyone or have we just been using it internally? It was added in #15781)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-:-KrWCy2p1pQfTWIn3Jic:b-xfj7ij)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -3548,6 +3543,27 @@ func (r *Replica) tickRaftMuLocked() (bool, error) {
+	} else if r.mu.quiescent {
+		done = true
+		if tickQuiesced {
+			// NB: It is safe to call TickQuiesced without holding Replica.raftMu
```
You couldn't have two goroutines trying to tick the replica at the same time and thus clobber the counter in undefined ways?

tschottdorf replied:
Ah, you hold the replica mu. Nevermind.

petermattis commented:
Removed the first commit, leaving just the fast-path for ticking.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/base/config.go, line 77 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-KrWCEaSzreyBWcjBZjq:-KrWNJjyYR-bWv8oOlpX:bx6s703) ([raw file](https://github.com/cockroachdb/cockroach/blob/36f9837c96863fde717b927fe908612607fc9334/pkg/base/config.go#L77)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This change is backwards-incompatible for anyone who might have been using this environment variable. (Have we recommended this to anyone or have we just been using it internally? It was added in #15781)
</blockquote></details>

Removed this commit from the PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-:-KrWNSiN1rlPrqADaCml:b7l5nqq)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 2 of 2 files at r3.
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/replica.go, line 3560 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-KrWRjmEYQZpibDzZGCl:-KrWRjmEYQZpibDzZGCm:bl2rbex) ([raw file](https://github.com/cockroachdb/cockroach/blob/48fe84dc916b6bba05c5ef52b051796c150fe195/pkg/storage/replica.go#L3560)):*
> ```Go
> 			// because that method simply increments a counter without performing any
> 			// other logic.
> 			r.mu.internalRaftGroup.TickQuiesced()
> ```

it's a shame that this same logic is in tickRaftMuLocked.

---

*[pkg/storage/store.go, line 3593 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-KrWR0H3qAmbBzJDW1hD:-KrWR0H3qAmbBzJDW1hE:blzce23) ([raw file](https://github.com/cockroachdb/cockroach/blob/48fe84dc916b6bba05c5ef52b051796c150fe195/pkg/storage/store.go#L3593)):*
> ```Go
> 				// disruption. Replica.maybeTickQuiesced only grabs short-duration
> 				// locks and not locks that are held during disk I/O.
> 				if r := v.(*Replica); r.maybeTickQuiesced() {
> ```

nit: what's the point of binding `r`? this looked like a checked type assertion to me briefly. could just be `if v.(*Replica).maybeTickQuiesced() {`

while we're at it, it's perhaps slightly clearer as:
```
if !v.(*Replica).maybeTickQuiesced() {
  rangeIDs = append(rangeIDs, k.(roachpb.RangeID))
}
return true
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-:-KrWROzgZd79lQFvisDb:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 3560 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-KrWRjmEYQZpibDzZGCl:-KrWTjT2PE1EAFoX8QRa:b7bybxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/48fe84dc916b6bba05c5ef52b051796c150fe195/pkg/storage/replica.go#L3560)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

it's a shame that this same logic is in tickRaftMuLocked.
</blockquote></details>

I feel no shame about it. Trying to share the code seems more trouble than its worth (IMO).

---

*[pkg/storage/store.go, line 3593 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617#-KrWR0H3qAmbBzJDW1hD:-KrWVlFIrzX2LMPLF5nz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/48fe84dc916b6bba05c5ef52b051796c150fe195/pkg/storage/store.go#L3593)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: what's the point of binding `r`? this looked like a checked type assertion to me briefly. could just be `if v.(*Replica).maybeTickQuiesced() {`

while we're at it, it's perhaps slightly clearer as:
```
if !v.(*Replica).maybeTickQuiesced() {
  rangeIDs = append(rangeIDs, k.(roachpb.RangeID))
}
return true
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17617)*
<!-- Sent from Reviewable.io -->
